### PR TITLE
add Scaled Exponential Linear Unit activation

### PR DESCRIPTION
**Pull request by rilut**
Thanks

### CODE REVIEW

**Comment by zafarali** [comment]
The paper also added a new dropout function which seems to be complentary to the performance. I don't think this should Be merged without that.

File: keras/activations.py [hunk 1]
```diff
@@ -34,6 +34,20 @@ def elu(x, alpha=1.0):
     return K.elu(x, alpha)
 
 
+def selu(x, alpha=1.6732632423543772848170429916717, scale=1.0507009873554804934193349852946):
```
Comment (fchollet) [inline keras/activations.py]: No point in providing those as arguments (we already have `elu` if you want custom values). They should be hard-coded.

**Comment by fchollet** [comment]
Also, should be unit-tested.

**Comment by rilut** [comment]
Thanks, I'll update my PR

**Comment by Dref360** [comment]
You already have a PR here : https://github.com/farizrahman4u/keras-contrib/pulls
We should not push the same feature at both place

**Comment by Dref360** [comment]
You already have a PR here : https://github.com/farizrahman4u/keras-contrib/pull/104
We should not push the same feature at both place

**Comment by rilut** [comment]
@Dref360 yes. after pushing to `keras`, I remember new layers and activations should go to `keras-contrib`, so I was going to close this PR. but looks like I'll continue here

**Comment by fchollet** [comment]
This one specific development is important enough that we can include it in core. It's not some random arxiv paper :)

File: keras/layers/core.py [hunk 2]
```diff
@@ -264,6 +264,61 @@ def _get_noise_shape(self, inputs):
         return noise_shape
 
 
+class AlphaDropout(Dropout):
```
Comment (fchollet) [inline keras/layers/core.py]: This should be in `noise`

File: keras/layers/core.py [hunk 3]
```diff
@@ -264,6 +264,61 @@ def _get_noise_shape(self, inputs):
...
+            def dropped_inputs(inputs=inputs, rate=self.rate, alpha=self.alpha,
+                               noise_shape=noise_shape, fixed_point_mean=self.fixed_point_mean,
+                               fixed_point_var=self.fixed_point_var, seed=self.seed):
+                retain_prob = 1.0 - rate
+                random_tensor = retain_prob
+                random_tensor += K.random_uniform(noise_shape, dtype=inputs.dtype, seed=seed)
+                binary_tensor = K.floor(random_tensor)
+
+                ret = inputs * binary_tensor + alpha * (1 - binary_tensor)
+                a = K.sqrt(fixed_point_var / (retain_prob * ((1 - retain_prob) * K.pow(alpha - fixed_point_mean, 2) + fixed_point_var)))
```
Comment (fchollet) [inline keras/layers/core.py]: Your code has super long lines. Please fix.

**Comment by rilut** [comment]
It's actually already green but I had to fix some typo and forget to add `[skip ci]`. Any more things I should've aware of?

Thanks

File: keras/activations.py [hunk 4]
```diff
@@ -34,6 +34,22 @@ def elu(x, alpha=1.0):
     return K.elu(x, alpha)
 
 
+def selu(x):
+    """Scaled Exponential Linear Unit. (Klambauer et al., 2017)
+
+    # Arguments
+        x: A tensor or variable to compute the activation function for.
+        alpha: A scalar, slope of positive section.
```
Comment (ihendley) [inline keras/activations.py]: Looks like alpha and scale are no longer arguments, so you can remove them from the docs here

Comment (rilut) [inline keras/activations.py]: @ihendley got it, thanks

**Comment by raghakot** [comment]
The paper also mentions a weight initialization scheme with mean = 0 and variance = sqrt(1/N).
he_normal has var = sqrt(2 / N). Should we be adding selu_normal init?
```
def selu_normal(seed=None):
    return VarianceScaling(scale=1.,
                           mode='fan_in',
                           distribution='normal',
                           seed=seed)
```

**Comment by rilut** [comment]
@raghakot good catch, thanks. I think that's important

File: keras/layers/noise.py [hunk 5]
```diff
@@ -90,3 +90,62 @@ def get_config(self):
...
+
+            def dropped_inputs(inputs=inputs, rate=self.rate, seed=self.seed):
+                alpha = 1.6732632423543772848170429916717
+                scale = 1.0507009873554804934193349852946
+                alpha_p = -alpha * scale
+
+                # get affine transformation params
+                a_eq = rate + K.pow(alpha_p, 2) * rate * (1 - rate)
+                a = K.pow(a_eq, -0.5)
+                b = -a * (alpha_p * (1 - rate))
```
Comment (pleax) [inline keras/layers/noise.py]: I believe this is incorrect implementation.

In the original paper `a` and `b` are defined as follows:
```tex
a = (q + \alpha' q (1 - q))^{-1/2}
b = -(a (1 - q) \alpha')
```
where `q` is keep probability. But here drop probability `rate` is used instead.

Comment (rilut) [inline keras/layers/noise.py]: @pleax thanks, this is very important

File: keras/layers/noise.py [hunk 6]
```diff
@@ -90,3 +90,64 @@ def get_config(self):
         config = {'rate': self.rate}
         base_config = super(GaussianDropout, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
+
+
+class AlphaDropout(Layer):
```
Comment (Dref360) [inline keras/layers/noise.py]: Is there a reason why there is not a `SpatialAlphaDropout` like there is a `SpatialDropout`?
In the paper they are not explicitly doing it, but they do have an argument `noise_shape` on their Github. 
When they release the code for more advanced datasets, we'll know for sure I guess.

Comment (rilut) [inline keras/layers/noise.py]: There was `noise_shape` when it was in Dropout/core layers. I'll just add it back, I guess

Comment (rilut) [inline keras/layers/noise.py]: added `SpatialAlphaDropout1D` class too

File: keras/layers/noise.py [hunk 7]
```diff
@@ -90,3 +91,103 @@ def get_config(self):
...
+            def dropped_inputs(inputs=inputs, rate=self.rate, seed=self.seed):
+                alpha = 1.6732632423543772848170429916717
+                scale = 1.0507009873554804934193349852946
+                alpha_p = -alpha * scale
+
+                kept_idx = K.greater_equal(K.random_uniform(noise_shape, seed=seed), rate)
+                kept_idx = K.cast(kept_idx, K.floatx())
+
+                # q is the `keep` probability
+                q = 1 - rate
```
Comment (fchollet) [inline keras/layers/noise.py]: Keeping `rate` does not make the expressions below more complicated; it makes them more readable:

```python
a = (1 - rate) * (1 + rate * alpha_p ** 2) ** (-0.5)
b = -a * (alpha_p * rate)
```

File: keras/layers/noise.py [hunk 8]
```diff
@@ -90,3 +91,103 @@ def get_config(self):
         config = {'rate': self.rate}
         base_config = super(GaussianDropout, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
+
+
+class AlphaDropout(Layer):
+    """Applies Alpha Dropout to the input.
+
+    Alpha Dropout is a Dropout that keeps mean and variance of inputs
```
Comment (fchollet) [inline keras/layers/noise.py]: Use ` around code keywords everywhere in docstrings..

File: keras/layers/noise.py [hunk 9]
```diff
@@ -90,3 +91,103 @@ def get_config(self):
...
+                x = inputs * kept_idx + alpha_p * (1 - kept_idx)
+
+                # Do affine transformation
+                return a * x + b
+
+            return K.in_train_phase(dropped_inputs, inputs, training=training)
+        return inputs
+
+
+class SpatialAlphaDropout1D(AlphaDropout):
```
Comment (fchollet) [inline keras/layers/noise.py]: Is this really necessary?

File: keras/layers/noise.py (hunk 6 above)
Comment (fchollet) [inline keras/layers/noise.py]: Is `AlphaDropout` really the canonical name for it? Let's make sure of it.

File: keras/initializers.py [hunk 10]
```diff
@@ -371,6 +371,28 @@ def he_normal(seed=None):
                            seed=seed)
 
 
+def selu_normal(seed=None):
```
Comment (fchollet) [inline keras/initializers.py]: Call it `lecun_normal` for consistency with `lecun_uniform`.

File: keras/layers/noise.py (hunk 7 above)
Comment (pleax) [inline keras/layers/noise.py]: Some parentheses missed, and some aren't required. I would propose:
```
- a = (1 - rate) * (1 + rate * alpha_p ** 2) ** (-0.5)
- b = -a * (alpha_p * rate)
+ a = ((1 - rate) * (1 + rate * alpha_p ** 2)) ** -0.5
+ b = -a * alpha_p * rate
```

File: keras/layers/noise.py (hunk 6 above)
Comment (zafarali) [inline keras/layers/noise.py]: pg 6 of the paper says:

> Therefore, we propose “alpha dropout”, that randomly sets inputs to α

File: keras/initializers.py (hunk 10 above)
Comment (zafarali) [inline keras/initializers.py]: Out of curiosity, why would you call it `lecun_normal`?

Comment (zafarali) [inline keras/initializers.py]: I think the question is more: it sounds like these authors have proposed the initialization scheme and therefore it should be called `klambauer_normal` or `hochreiter_normal`?

File: tests/keras/layers/noise_test.py [hunk 11]
```diff
@@ -23,5 +23,14 @@ def test_GaussianDropout():
                input_shape=(3, 2, 3))
 
 
+@keras_test
+@pytest.mark.skipif((K.backend() == 'cntk'),
+                    reason="cntk does not support it yet")
```
Comment (arodiss) [inline tests/keras/layers/noise_test.py]: Would be nice to replace "it" with more exact reference

Comment (rilut) [inline tests/keras/layers/noise_test.py]: nice, but I'm not so sure about this

Comment (rilut) [inline tests/keras/layers/noise_test.py]: nice, but I don't know what prevents this

Comment (arodiss) [inline tests/keras/layers/noise_test.py]: I mean to replace "cntk does not support it yet" with eg "cntk does not support AlphaDropout yet"
This way whenever I see this notice I can check whether or not it is still the case - and hopefully test will be unskipped as soon as CNTK comes up with necessary feature

File: keras/layers/noise.py [hunk 12]
```diff
@@ -153,3 +153,8 @@ def dropped_inputs(inputs=inputs, rate=self.rate, seed=self.seed):
 
             return K.in_train_phase(dropped_inputs, inputs, training=training)
         return inputs
+
+    def get_config(self):
```
Comment (zafarali) [inline keras/layers/noise.py]: 👍

File: keras/initializers.py (hunk 10 above)
Comment (fchollet) [inline keras/initializers.py]: LeCun proposed this scheme way back (see ref in `lecun_uniform` docstring). Our initializers are named `*_normal` and `*_uniform`. We already have the `lecun_uniform` initializer, and this initializer is simply the normal version of it. I don't even know why didn't already have it, it's a significant inconsistency.

**Comment by bigsnarfdude** [comment]
I tried to make some examples to understand usage of SELU and AlphaDropout. I'm not confident that I'm implemented it correctly. I would like to take a stab at CIFAR10 example if I can get the basics correct here in this repo: https://github.com/bigsnarfdude/SELU_Keras_Tutorial

**Comment by zafarali** [comment]
I think it's still unclear how SELU performs in CNN settings. The authors suggest that it is awesome for MLPs, therefore an MLP example would be good (you have one in your repo @bigsnarfdude)

**Comment by zafarali** [comment]
If not I can modify this and make a PR once this is merged: https://github.com/fchollet/keras/blob/master/examples/reuters_mlp.py

File: keras/initializers.py [hunk 13]
```diff
@@ -371,7 +371,7 @@ def he_normal(seed=None):
                            seed=seed)
 
 
-def selu_normal(seed=None):
+def lecun_normal(seed=None):
     """SELU normal initializer.
```
Comment (pleax) [inline keras/initializers.py]: Docstring

**Comment by fchollet** [comment]
LGTM, thanks
